JIT: Intrinsify UTF16->UTF8 conversion for string literals (Encoding.UTF8.TryGetBytes)

EgorBo opened the pull request:
Closes https://github.com/dotnet/runtime/issues/84659
The main goal is to make upcoming UTF8 string interpolation faster (to be on par with UTF16 one). If JIT sees that the source is a string literal (can be extended to cover RVA `ROS<char>`, "shifted" pointers to literal literals, `static readonly string`) and the data in it is ASCII only - it peforms the conversion in the JIT time and just copies it to the destination.

## Benchmark:

```csharp
byte[] _outputBuffer = new byte[1024];

[Benchmark]
public int TryGetBytes_const()
{
    Encoding.UTF8.TryGetBytes("hello", _outputBuffer.AsSpan(), out int written1);
    return written1;
}
```

|            Method |                   Toolchain |      Mean | Ratio |
|------------------ |---------------------------- |----------:|------:|
| TryGetBytes_const |      \Core_Root\corerun.exe | **0.3435 ns** |  0.06 |
| TryGetBytes_const | \Core_Root_base\corerun.exe | 5.4439 ns |  1.00 |

cc @stephentoub

EgorBo commented:
@jakobbotsch @SingleAccretion yet another `gtSplitTree` use case PTAL 🙂 not urgent

I recommend to enable "ignore whitespaces" mode there were a couple of "refactorings" + a pretty simple expansion for `GetUtf8Bytes`

jakobbotsch commented:
@EgorBo Can you merge and rerun CI?

jakobbotsch commented on src/coreclr/jit/compiler.h:
```diff
@@ -8873,6 +8891,42 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+                return TYP_INT;
+            case 8:
+                return TYP_LONG;
+#ifdef FEATURE_SIMD
+            case 16:
+                return TYP_SIMD16;
+#ifdef TARGET_XARCH
+            case 32:
+                return TYP_SIMD32;
+            case 64:
+                return TYP_SIMD64;
+#endif
+#endif
```
A question (I also have some similar logic in #85323) -- is it ok to create nodes with these SIMD16/32/64 types without checking for support for these types in some way?

EgorBo replied:
In this case support is checked, this method invokes `roundDownSIMDSize` that does that

jakobbotsch commented on src/coreclr/jit/compiler.h:
```diff
@@ -7112,6 +7120,16 @@ class Compiler
         optMethodFlags |= OMF_HAS_TLS_FIELD;
     }
 
+    bool doesMethodHasSpecialIntrinsics()
```
```suggestion
    bool doesMethodHaveSpecialIntrinsics()
```

jakobbotsch commented on src/coreclr/jit/helperexpansion.cpp:
```diff
@@ -413,6 +413,388 @@ bool Compiler::fgExpandRuntimeLookupsForCall(BasicBlock** pBlock, Statement* stm
+{
+    PhaseStatus result = PhaseStatus::MODIFIED_NOTHING;
+
+    if (!doesMethodHasSpecialIntrinsics() || opts.OptimizationDisabled())
```
```suggestion
    if (!doesMethodHaveSpecialIntrinsics() || opts.OptimizationDisabled())
```

jakobbotsch commented on src/coreclr/jit/helperexpansion.cpp:
```diff
@@ -413,6 +413,388 @@ bool Compiler::fgExpandRuntimeLookupsForCall(BasicBlock** pBlock, Statement* stm
+
+    // Read the string literal (UTF16) into a local buffer (UTF8)
+    assert(strObj != nullptr);
+    BYTE srcUtf8cns[MaxPossibleUnrollThreshold] = {}; // same length since we're narrowing U16 to U8
```
```suggestion
    BYTE srcUtf8cns[MaxPossibleUnrollThreshold]; // same length since we're narrowing U16 to U8
```

Is there a reason to zero these 256 bytes?

jakobbotsch commented on src/coreclr/jit/helperexpansion.cpp:
```diff
@@ -413,6 +413,388 @@ bool Compiler::fgExpandRuntimeLookupsForCall(BasicBlock** pBlock, Statement* stm
+    for (unsigned charIndex = 0; charIndex < srcLenCns; charIndex++)
+    {
+        uint16_t ch = 0;
+        if (!info.compCompHnd->getStringChar(strObj, (int)charIndex, &ch))
```
Seems unfortunate that we have to do a JIT-EE call for every single character. Didn't you add more general purpose API that could be used to bulk copy this data?

jakobbotsch replied:
In SPMI every single of these queries is going to cost us 24+ bytes in storage.

EgorBo replied:
Yes, we have `getObjectContent` now

jakobbotsch commented on src/coreclr/jit/helperexpansion.cpp:
```diff
@@ -413,6 +413,388 @@ bool Compiler::fgExpandRuntimeLookupsForCall(BasicBlock** pBlock, Statement* stm
+    // srcLenCns is the length of the string literal in chars (UTF16)
+    // but we're going to use the same value as the "bytesWritten" result in the fast path and in the length check.
+    GenTree* srcLenCnsNode = gtNewIconNode(srcLenCns);
+    fgUpdateConstTreeValueNumber(srcLenCnsNode);
```
```suggestion
    fgValueNumberTreeConst(srcLenCnsNode);
```

jakobbotsch commented on src/coreclr/jit/helperexpansion.cpp:
```diff
@@ -413,6 +413,388 @@ bool Compiler::fgExpandRuntimeLookupsForCall(BasicBlock** pBlock, Statement* stm
+    const bool thisIsKnownNonNull =
+        thisArg->gtVNPair.BothEqual() && vnStore->IsKnownNonNull(thisArg->gtVNPair.GetLiberal());
+
+    // Spill all original arguments to locals in the lengthCheckBb to preserve all possible side-effects.
+    thisArg = SpillExpression(this, thisArg, lengthCheckBb, debugInfo);
+    srcPtr  = SpillExpression(this, srcPtr, lengthCheckBb, debugInfo);
+    srcLen  = SpillExpression(this, srcLen, lengthCheckBb, debugInfo);
+    dstPtr  = SpillExpression(this, dstPtr, lengthCheckBb, debugInfo);
+    dstLen  = SpillExpression(this, dstLen, lengthCheckBb, debugInfo);
```
Is this necessary? These should have been spilled by `gtSplitTree`.

jakobbotsch replied:
These node pointers are from before `gtSplitTree` made changes, it seems very dangerous to rely on them. You should rather reload them from the individual `CallArg`s.

jakobbotsch commented on src/coreclr/jit/helperexpansion.cpp:
```diff
@@ -413,6 +413,388 @@ bool Compiler::fgExpandRuntimeLookupsForCall(BasicBlock** pBlock, Statement* stm
+    if (!noSizeCheck)
+    {
+        // Set bytesWritten -1 by default, if the fast path is not taken we'll return it as the result.
+        GenTree* bytesWrittenDefaultVal = gtNewAssignNode(gtClone(resultLcl), gtNewIconNode(-1));
```
```suggestion
        GenTree* bytesWrittenDefaultVal = gtNewAssignNode(gtCloneExpr(resultLcl), gtNewIconNode(-1));
```

No reason to use `gtClone` over `gtCloneExpr` if you aren't using it for the fact that it can fail for complex trees, IMO.

jakobbotsch replied:
This will conflict with #85585 I think.

jakobbotsch replied:
I just merged #85585 so I think you'll need to adapt this a bit.

SingleAccretion replied:
For reference, methods for creating store nodes:
1) `gtNewStoreLclVarNode` - former `ASG(LCL_VAR, ...)`.
2) `gtNewStoreLclFldNode` - former `ASG(LCL_FLD, ...)`.
3) `gtNewStoreIndNode` - former `ASG(IND, ...)`.
4) `gtNewStoreBlkNode` - former `ASG(BLK, ...)`.

EgorBo replied:
thanks! Yeah, I follow your PRs so I'm aware 🙂

jakobbotsch commented on src/coreclr/jit/helperexpansion.cpp:
```diff
@@ -413,6 +413,388 @@ bool Compiler::fgExpandRuntimeLookupsForCall(BasicBlock** pBlock, Statement* stm
+    // (after we evaluated all arguments)
+    if (!thisIsKnownNonNull)
+    {
+        GenTree* thisNullcheck = gtNewNullCheck(gtClone(thisArg), lengthCheckBb);
```
```suggestion
        GenTree* thisNullcheck = gtNewNullCheck(gtCloneExpr(thisArg), lengthCheckBb);
```

jakobbotsch commented on src/coreclr/jit/helperexpansion.cpp:
```diff
@@ -413,6 +413,388 @@ bool Compiler::fgExpandRuntimeLookupsForCall(BasicBlock** pBlock, Statement* stm
+
+    if (!noSizeCheck)
+    {
+        GenTree* lengthCheck = gtNewOperNode(GT_LT, TYP_INT, gtClone(dstLen), srcLenCnsNode);
```
```suggestion
        GenTree* lengthCheck = gtNewOperNode(GT_LT, TYP_INT, gtCloneExpr(dstLen), srcLenCnsNode);
```

jakobbotsch commented on src/coreclr/jit/helperexpansion.cpp:
```diff
@@ -413,6 +413,388 @@ bool Compiler::fgExpandRuntimeLookupsForCall(BasicBlock** pBlock, Statement* stm
+    // The widest type we can use for loads
+    const var_types maxLoadType = roundDownMaxRegSize(srcLenCns);
+
+    // How many iterations we need to copy UTF8 const data to the destination
+    unsigned iterations = srcLenCns / genTypeSize(maxLoadType);
+
+    // Add one more iteration if we have a remainder
+    iterations += srcLenCns % genTypeSize(maxLoadType) == 0 ? 0 : 1;
```
```suggestion
    // How many iterations we need to copy UTF8 const data to the destination
    unsigned iterations = srcLenCns / genTypeSize(maxLoadType);

    // Add one more iteration if we have a remainder
    iterations += (srcLenCns % genTypeSize(maxLoadType)) == 0 ? 0 : 1;
```

(or write it in the `(x + (y - 1) ) / y` way, that's usually how we do it in the JIT I think)

jakobbotsch commented on src/coreclr/jit/helperexpansion.cpp:
```diff
@@ -413,6 +413,388 @@ bool Compiler::fgExpandRuntimeLookupsForCall(BasicBlock** pBlock, Statement* stm
+
+    for (unsigned i = 0; i < iterations; i++)
+    {
+        ssize_t offest = (ssize_t)i * genTypeSize(maxLoadType);
```
```suggestion
        ssize_t offset = (ssize_t)i * genTypeSize(maxLoadType);
```

jakobbotsch commented on src/coreclr/jit/helperexpansion.cpp:
```diff
@@ -413,6 +413,388 @@ bool Compiler::fgExpandRuntimeLookupsForCall(BasicBlock** pBlock, Statement* stm
+        // Last iteration: overlap with previous load if needed
+        if (i == iterations - 1)
+        {
+            offest = (ssize_t)srcLenCns - genTypeSize(maxLoadType);
```
```suggestion
            offset = (ssize_t)srcLenCns - genTypeSize(maxLoadType);
```

jakobbotsch commented on src/coreclr/jit/helperexpansion.cpp:
```diff
@@ -413,6 +413,388 @@ bool Compiler::fgExpandRuntimeLookupsForCall(BasicBlock** pBlock, Statement* stm
+        // -----------       |     \--*  CNS_INT   int   %offset%
+        // -----------       \--*  CNS_VEC or CNS_INT representing UTF8 const data chunk
+
+        GenTreeIntCon* offsetNode = gtNewIconNode(offest);
```
```suggestion
        GenTreeIntCon* offsetNode = gtNewIconNode(offset);
```

jakobbotsch commented on src/coreclr/jit/helperexpansion.cpp:
```diff
@@ -413,6 +413,388 @@ bool Compiler::fgExpandRuntimeLookupsForCall(BasicBlock** pBlock, Statement* stm
+        // -----------       \--*  CNS_VEC or CNS_INT representing UTF8 const data chunk
+
+        GenTreeIntCon* offsetNode = gtNewIconNode(offest);
+        fgUpdateConstTreeValueNumber(offsetNode);
```
```suggestion
        fgValueNumberTreeConst(offsetNode);
```

jakobbotsch commented on src/coreclr/jit/helperexpansion.cpp:
```diff
@@ -413,6 +413,388 @@ bool Compiler::fgExpandRuntimeLookupsForCall(BasicBlock** pBlock, Statement* stm
+
+        // Grab a chunk from srcUtf8cnsData for the given offset and width
+        GenTree* utf8cnsChunkNode = gtNewCon(maxLoadType, srcUtf8cns + offest);
+        fgUpdateConstTreeValueNumber(utf8cnsChunkNode);
```
```suggestion
        fgValueNumberTreeConst(utf8cnsChunkNode);
```

jakobbotsch commented on src/coreclr/jit/helperexpansion.cpp:
```diff
@@ -413,6 +413,388 @@ bool Compiler::fgExpandRuntimeLookupsForCall(BasicBlock** pBlock, Statement* stm
+        fgUpdateConstTreeValueNumber(offsetNode);
+
+        // Grab a chunk from srcUtf8cnsData for the given offset and width
+        GenTree* utf8cnsChunkNode = gtNewCon(maxLoadType, srcUtf8cns + offest);
```
```suggestion
        GenTree* utf8cnsChunkNode = gtNewCon(maxLoadType, srcUtf8cns + offset);
```

jakobbotsch commented on src/coreclr/jit/helperexpansion.cpp:
```diff
@@ -413,6 +413,388 @@ bool Compiler::fgExpandRuntimeLookupsForCall(BasicBlock** pBlock, Statement* stm
+        GenTree* utf8cnsChunkNode = gtNewCon(maxLoadType, srcUtf8cns + offest);
+        fgUpdateConstTreeValueNumber(utf8cnsChunkNode);
+
+        GenTree*   dstAddOffsetNode = gtNewOperNode(GT_ADD, TYP_BYREF, gtClone(dstPtr), offsetNode);
```
```suggestion
        GenTree*   dstAddOffsetNode = gtNewOperNode(GT_ADD, TYP_BYREF, gtCloneExpr(dstPtr), offsetNode);
```

jakobbotsch commented on src/coreclr/jit/helperexpansion.cpp:
```diff
@@ -413,6 +413,388 @@ bool Compiler::fgExpandRuntimeLookupsForCall(BasicBlock** pBlock, Statement* stm
+        GenTreeOp* storeInd         = gtNewAssignNode(indirNode, utf8cnsChunkNode);
+        Statement* storeIndStmt     = fgNewStmtFromTree(storeInd, debugInfo);
+        fgInsertStmtAtEnd(fastpathBb, storeIndStmt);
+        gtUpdateStmtSideEffects(storeIndStmt);
```
```suggestion
```

Don't think this is necessary, the nodes should propagate flags as they are created.

jakobbotsch commented on src/coreclr/jit/helperexpansion.cpp:
```diff
@@ -413,6 +413,388 @@ bool Compiler::fgExpandRuntimeLookupsForCall(BasicBlock** pBlock, Statement* stm
+
+    // Finally, store the number of bytes written to the resultLcl local
+    Statement* finalStmt =
+        fgNewStmtFromTree(gtNewAssignNode(gtClone(resultLcl), gtCloneExpr(srcLenCnsNode)), debugInfo);
```
```suggestion
        fgNewStmtFromTree(gtNewAssignNode(gtCloneExpr(resultLcl), gtCloneExpr(srcLenCnsNode)), debugInfo);
```

jakobbotsch commented on src/coreclr/jit/helperexpansion.cpp:
```diff
@@ -764,7 +1146,14 @@ bool Compiler::fgExpandHelperForBlock(BasicBlock** pBlock)
-            if (!tree->IsHelperCall())
+            if (handleIntrinsics)
+            {
+                if (!tree->IsCall() || !(tree->AsCall()->gtCallMoreFlags & GTF_CALL_M_SPECIAL_INTRINSIC))
```
```suggestion
                if (!tree->IsCall() || ((tree->AsCall()->gtCallMoreFlags & GTF_CALL_M_SPECIAL_INTRINSIC) == 0))
```

According to our conventions.

jakobbotsch commented on src/coreclr/jit/morph.cpp:
```diff
@@ -7754,6 +7754,16 @@ GenTree* Compiler::fgMorphCall(GenTreeCall* call)
 #endif
     }
 
+    if (call->gtCallMoreFlags & GTF_CALL_M_SPECIAL_INTRINSIC)
```
```suggestion
    if ((call->gtCallMoreFlags & GTF_CALL_M_SPECIAL_INTRINSIC) != 0)
```

jakobbotsch commented on src/coreclr/jit/morph.cpp:
```diff
@@ -7754,6 +7754,16 @@ GenTree* Compiler::fgMorphCall(GenTreeCall* call)
 
+    if (call->gtCallMoreFlags & GTF_CALL_M_SPECIAL_INTRINSIC)
+    {
+        if (lookupNamedIntrinsic(call->AsCall()->gtCallMethHnd) ==
+            NI_System_Text_UTF8Encoding_UTF8EncodingSealed_GetUtf8Bytes)
+        {
+            // Expanded in fgVNBasedIntrinsicExpansion
+            setMethodHasSpecialIntrinsics();
+        }
```
Why set this in morph instead of during import where we already have the named intrinsic available? The flag name also seems inaccurate, maybe rename it to `MethodHasExpandableSpecialIntrinsic`?

EgorBo replied:
I usually do this in morph because in importer we would need to transfer this flag from inlinee's compiler to the root one (or we can always set it in root's one but that is a bit conservative, we have a few cases where we reject inlining on the go, e.g. for generics with runtime lookups)

jkotas commented on src/coreclr/jit/importercalls.cpp:
```diff
@@ -8540,6 +8541,17 @@ NamedIntrinsic Compiler::lookupNamedIntrinsic(CORINFO_METHOD_HANDLE method)
             }
+            else if (strcmp(namespaceName, "Text") == 0)
+            {
+                if (strcmp(className, "UTF8EncodingSealed") == 0)
```
For the UTF8 string interpolation, would it be better to just have a simple static `UTF8.TryGetBytes` method that would be easier for the JIT to deal with?

It feels like the JIT spends a lot of effort to reverse engineer the pattern. It is likely going to eat into inlining budgets and other limitations. How complicated the interpolated string can be before the optimizations get turned off? 

cc @stephentoub

stephentoub replied:
If it's really a problem, we could at a minimum add that as an internal method that Utf8.TryWrite would use.  I'm not super excited about adding two ways to do the exact same thing publicly, i.e. Utf8.TryGetBytes vs Encoding.UTF8.TryGetBytes, though I also don't love the prospect of having the special one we use internally in corelib and a public one available for everything else, so ideally this would only be a medium-term mitigation.

EgorBo replied:
Initially, I planned to intrinsify the public API itself `bool TryGetBytes(ReadOnlySpan<char> chars, Span<byte> bytes, out int bytesWritten)` and even made a prototype for. I then realized that Span arguments are very hard to handle in JIT because the way we do ABI preparations too early, I migth try again that.

But an api that doesn't take spans as arguments is a lot easier to work with

EgorBo replied:
Let me try to handle `TryGetBytes` itself. It is currently never inlined without Dynamic PGO so I won't regress anyone by adding a NoInlining attribute for it. (well, except PGO users)

EgorBo replied:
> > But an api that doesn't take spans as arguments is a lot easier to work with
> 
> So something like `private static bool TryGetBytes(ref char source, int sourceLength, ref byte destination, int destinationLength)`?

Yes, but let me see if Span can work too again

BruceForstall commented:
@EgorBo Do you want to mark this as "Draft" so it doesn't get considered as a stale PR?

EgorBo committed:
Revert to the previous version and address Jakob's feedback

EgorBo committed:
Clean up

EgorBo committed:
Fix test

EgorBo committed:
fix managed api

EgorBo committed:
fix test

EgorBo committed:
Merge branch 'main' of github.com:dotnet/runtime into u16-u8-better

EgorBo committed:
Clean up

EgorBo commented:
```csharp
byte[] _outputBuffer = new byte[1024];

[Benchmark]
public int TryGetBytes_const()
{
    Encoding.UTF8.TryGetBytes("hello", _outputBuffer.AsSpan(), out int written1);
    return written1;
}

[Benchmark]
[Arguments("hello")]
public int TryGetBytes_var(string s1)
{
    Encoding.UTF8.TryGetBytes(s1, _outputBuffer.AsSpan(), out int written1);
    return written1;
}
```

|            Method |                   Toolchain |      Mean | Ratio |
|------------------ |---------------------------- |----------:|------:|
| TryGetBytes_const |      \Core_Root\corerun.exe | 0.3435 ns |  0.06 |
| TryGetBytes_const | \Core_Root_base\corerun.exe | 5.4439 ns |  1.00 |
|                   |                             |           |       |
|   TryGetBytes_var |      \Core_Root\corerun.exe | 4.7716 ns |  1.12 |
|   TryGetBytes_var | \Core_Root_base\corerun.exe | 4.2483 ns |  1.00 |

More complicated benchmark, larger data:
```cs
byte[] _outputBuffer = new byte[1024];

[Benchmark]
public int TryGetBytes_const()
{
    Encoding.UTF8.TryGetBytes("The MIT License (MIT)", _outputBuffer.AsSpan(), out int written1);
    Encoding.UTF8.TryGetBytes("Copyright (c) .NET Foundation and Contributors", _outputBuffer.AsSpan(written1), out int written2);
    Encoding.UTF8.TryGetBytes("All rights reserved.", _outputBuffer.AsSpan(written2), out int written3);
    return written1 + written2 + written3;
}

[Benchmark]
[Arguments(
    "The MIT License (MIT)", 
    "Copyright (c) .NET Foundation and Contributors",
    "All rights reserved.")]
public int TryGetBytes_var(string s1, string s2, string s3)
{
    Encoding.UTF8.TryGetBytes(s1, _outputBuffer.AsSpan(), out int written1);
    Encoding.UTF8.TryGetBytes(s2, _outputBuffer.AsSpan(written1), out int written2);
    Encoding.UTF8.TryGetBytes(s3, _outputBuffer.AsSpan(written2), out int written3);
    return written1 + written2 + written3;
}
```

|            Method |                   Toolchain |      Mean | Ratio |
|------------------ |---------------------------- |----------:|------:|
| TryGetBytes_const |      \Core_Root\corerun.exe |  2.677 ns |  0.16 |
| TryGetBytes_const | \Core_Root_base\corerun.exe | 16.373 ns |  1.00 |
|                   |                             |           |       |
|   TryGetBytes_var |      \Core_Root\corerun.exe | 19.149 ns |  1.09 |
|   TryGetBytes_var | \Core_Root_base\corerun.exe | 17.571 ns |  1.00 |


So we do receieve a relatively small penalty from a non-inline method we introduced (it should not regress users since the API is new). Just like @jkotas proposed I intrinsified a static method that can be e.g. moved to Utf8 class if needed and the Utf8 interpolation can call it **directly** if needed - I wired it up in `TryGetBytes` to also handle constant input for it as well, the penalty can be mitigated if we slightly improve `IsKnownConstant` in jit and will be able to detect cases where span points to a literal in advance, but feel free to suggest to unwire it if calling the static method directly is fine.

Also, addressed all @jakobbotsch feedback

stephentoub commented:
> the penalty can be mitigated if we slightly improve IsKnownConstant in jit and will be able to detect cases where span points to a literal in advance

Is this complicated / expensive?  If not, can we do that? I'd expect the common case for direct end-user usage of Encoding.UTF8.TryGetBytes to be with non-const inputs.  The const input case I've been primarily interested in is with Utf8.TryWrite and the AppendLiteral method that'll be called by code emitted by the C# compiler for the literal portion of interpolated strings.  If it's too complicated / expensive / etc. to avoid the overhead for non-const inputs with TryGetBytes, then we should just skip TryGetBytes altogether and have the Utf8.TryWriteInterpolatedStringHandler.AppendLiteral use this intrinsic directly.

jakobbotsch commented on src/coreclr/jit/helperexpansion.cpp:
```diff
@@ -868,7 +868,14 @@ bool Compiler::fgExpandHelperForBlock(BasicBlock** pBlock)
         for (GenTree* const tree : stmt->TreeList())
         {
-            if (!tree->IsHelperCall())
+            if (handleIntrinsics)
+            {
+                if (!tree->IsCall() || ((tree->AsCall()->gtCallMoreFlags & GTF_CALL_M_SPECIAL_INTRINSIC) == 0))
+                {
+                    continue;
+                }
+            }
+            else if (!tree->IsHelperCall())
```
Seems like it would be cleaner to leave this up to the expansion function and avoid the `handleIntrinsics` argument. The expansion function presumably has to do a bunch of call related checks anyway, so it can check for `IsHelperCall` or special intrinsic as necessary.

jakobbotsch commented on src/coreclr/jit/helperexpansion.cpp:
```diff
@@ -1171,3 +1178,331 @@ bool Compiler::fgExpandStaticInitForCall(BasicBlock** pBlock, Statement* stmt, G
+//
+bool Compiler::fgVNBasedIntrinsicExpansionForCall(BasicBlock** pBlock, Statement* stmt, GenTreeCall* call)
+{
+    assert(call->gtCallMoreFlags & GTF_CALL_M_SPECIAL_INTRINSIC);
```
```suggestion
    assert((call->gtCallMoreFlags & GTF_CALL_M_SPECIAL_INTRINSIC) != 0);
```

according to coding conventions

jakobbotsch commented on src/coreclr/jit/helperexpansion.cpp:
```diff
@@ -1171,3 +1178,331 @@ bool Compiler::fgExpandStaticInitForCall(BasicBlock** pBlock, Statement* stmt, G
+        GenTree* utf8cnsChunkNode = gtNewGenericCon(maxLoadType, buffer + offset);
+        fgValueNumberTreeConst(utf8cnsChunkNode);
+
+        GenTree*   dstAddOffsetNode = gtNewOperNode(GT_ADD, TYP_BYREF, gtCloneExpr(dstPtr), offsetNode);
```
```suggestion
        GenTree*   dstAddOffsetNode = gtNewOperNode(GT_ADD, dstPtr->TypeGet(), gtCloneExpr(dstPtr), offsetNode);
```

jakobbotsch commented on src/coreclr/jit/helperexpansion.cpp:
```diff
@@ -1171,3 +1178,331 @@ bool Compiler::fgExpandStaticInitForCall(BasicBlock** pBlock, Statement* stmt, G
+    for (unsigned charIndex = 0; charIndex < srcLenCns; charIndex++)
+    {
+        // Buffer keeps the original utf16 chars
+        uint16_t ch = ((uint16_t*)buffer)[charIndex];
```
Interpreting this buffer as both `BYTE` and `uint16_t` with pointer casts makes me feel a bit uneasy... maybe `memcpy` this?

EgorBo replied:
Thanks, decided to introduce a separate buffer for output - 256 bytes won't hurt it

EgorBo committed:
Apply suggestions from code review

Co-authored-by: Jakob Botsch Nielsen <Jakob.botsch.nielsen@gmail.com>

EgorBo committed:
Merge branch 'main' of github.com:dotnet/runtime into intrinsify-utf16-utf8-literal

EgorBo committed:
Address feedback

jakobbotsch commented on src/coreclr/jit/helperexpansion.cpp:
```diff
@@ -874,7 +873,12 @@ bool Compiler::fgExpandHelperForBlock(BasicBlock** pBlock)
+                continue;
+            }
+
+            if (!tree->IsHelperCall() && ((tree->AsCall()->gtCallMoreFlags & GTF_CALL_M_SPECIAL_INTRINSIC) == 0))
             {
                 continue;
             }
```
Why this check? Just for TP purposes? Does it actually make a difference?

EgorBo replied:
You're right - it's no longer needed

jakobbotsch reviewed (approved):
The JIT changes look good to me.

EgorBo committed:
Update helperexpansion.cpp

EgorBo commented:
> > the penalty can be mitigated if we slightly improve IsKnownConstant in jit and will be able to detect cases where span points to a literal in advance
> 
> Is this complicated / expensive? If not, can we do that? I'd expect the common case for direct end-user usage of Encoding.UTF8.TryGetBytes to be with non-const inputs. The const input case I've been primarily interested in is with Utf8.TryWrite and the AppendLiteral method that'll be called by code emitted by the C# compiler for the literal portion of interpolated strings. If it's too complicated / expensive / etc. to avoid the overhead for non-const inputs with TryGetBytes, then we should just skip TryGetBytes altogether and have the Utf8.TryWriteInterpolatedStringHandler.AppendLiteral use this intrinsic directly.

@stephentoub It's tricky to implement but it should be doable. Calling this static method directly from interpolation should be a better solution because we won't have to rely on inlining just like Jan pointed out earlier, so I guess for now I'll remove its usage from `TryGetValue` till IsKnonwConstant is implemented

EgorBo committed:
Merge branch 'main' of github.com:dotnet/runtime into intrinsify-utf16-utf8-literal

EgorBo committed:
Merge branch 'intrinsify-utf16-utf8-literal' of github.com:EgorBo/runtime-1 into intrinsify-utf16-utf8-literal

stephentoub commented:
> It's tricky to implement but it should be doable. Calling this static method directly from interpolation should be a better solution because we won't have to rely on inlining just like Jan pointed out earlier, so I guess for now I'll remove its usage from TryGetValue till IsKnonwConstant is implemented

Ok, sounds good.

EgorBo committed:
Merge branch 'main' of github.com:dotnet/runtime into intrinsify-utf16-utf8-literal

EgorBo commented:
/azp run runtime-coreclr pgo

stephentoub commented:
@EgorBo, anything I can do to help land this?

EgorBo committed:
Update UTF8Encoding.Sealed.cs

EgorBo commented:
> @EgorBo, anything I can do to help land this?

I was thinking of filing a separate PR to improve `IsKnownConstant` first, but I think it can be landed as is too in case if you want to call this new static `ReadUtf8` directly from the interpolation?

stephentoub commented:
> if you want to call this new static ReadUtf8 directly from the interpolation

Yes, I thought that's what we agreed on above, using it from AppendLiteral and not from TryGetBytes yet.

EgorBo commented:
Merging then, I've verified that the API works as expected when it's used inside `TryGetBytes` and removed it from there so now this API is unused. The additional runtime tests for `TryGetBytes` won't hurt I guess.

stephentoub commented:
I'll submit a pr. Thanks.

teo-tsirpanis commented on src/coreclr/jit/helperexpansion.cpp:
```diff
@@ -1177,3 +1179,336 @@ bool Compiler::fgExpandStaticInitForCall(BasicBlock** pBlock, Statement* stmt, G
+        uint16_t ch = bufferU16[charIndex];
+        if (ch > 127)
+        {
+            // Only ASCII is supported.
```
Why?
